plugin/segment/issue.go

```diff
@@ -0,0 +1,87 @@
+package segment
+
+import (
+	"context"
+	"time"
+
+	"github.com/bytebase/bytebase/api"
+	"github.com/bytebase/bytebase/store"
+	"github.com/segmentio/analytics-go"
+)
+
+// IssueReporter is the segment reporter for issue.
+type IssueReporter struct {
+}
+
+var issueTypeList = []api.IssueType{
+	api.IssueGeneral,
+	api.IssueDatabaseCreate,
+	api.IssueDatabaseGrant,
+	api.IssueDatabaseSchemaUpdate,
+	api.IssueDatabaseSchemaUpdate,
+	api.IssueDatabaseSchemaUpdateGhost,
+	api.IssueDatabaseDataUpdate,
+	api.IssueDataSourceRequest,
+}
+
+// Report will exec the segment reporter for issue
+func (t *IssueReporter) Report(ctx context.Context, store *store.Store, segment *segment) error {
+	total, err := store.CountIssue(ctx, &api.IssueFind{})
```

Comment:
The interface can be more general.

1. We don't need to pass in store, instead we define a MetricCollector interface and store just happens to implement that. So what to be collected is the store's implementation details. It also makes testing easier, since we can pass in a fake collector

2. The reporter should be decoupled from the collector. The segment implementation is a reporter, while the store implementation is a collector. Thus in our run method:

for {
 metrics := collector.collect()
 reporters.report(metrics)
}

Thus we can pass fake collector / reporter for testing as well.

Reply:
Also for each metric, like total issue, we should also define a metric name. e.g. bb.issue.count and in the store  init method:

collectorList.add ("bb.issue.count" , () {
   total, err := store.CountIssue(ctx, &api.IssueFind{})
  xxx
}

}

collect() {
  for c := collectorList {
   c.fn()  
  }
}

By doing this, we have a single place to see all metrics collected by the store.